sklearn/ensemble/ensemble_selection.py

```diff
@@ -0,0 +1,261 @@
+        ens_proba = np.zeros((X.shape[0], self.n_classes_))
+
+        for estimator_name in self.ensemble_:
+            weight = self.ensemble_[estimator_name]
+            proba = self.named_estimators[estimator_name].predict_proba(X)
+            ens_proba += proba * weight
+
+        ens_proba /= n_clfs
+        return np.argmax(ens_proba, axis=1)
+
+    # TODO:
+    #def predict_proba(self, X):
+
+
+    def _get_score_of_model(self, X, y, estimator):
+        proba = estimator.predict_proba(X)
+        score = self.score_metric_(y, proba)
+        return score, proba
+
+    def _get_score_of_ensemble(self, X, y, ensemble):
+        n_clfs = sum(ensemble.values())
+        ens_proba = np.zeros((X.shape[0], self.n_classes_))
+
+        for estimator_name in ensemble:
+            weight = ensemble[estimator_name]
+            proba = self.clfs_proba_[estimator_name]
+            ens_proba += proba * weight
+
+        ens_proba /= n_clfs
+        ens_score = self.score_metric_(y, ens_proba)
+        return ens_score, ens_proba
+
+    def _build_ensemble(self, X, y, ranked_estimators_name):
+        ensemble = Counter(ranked_estimators_name[:self.n_best])
+        ens_score, ens_proba = self._get_score_of_ensemble(X, y, ensemble)
+        n_clfs = sum(ensemble.values())
+
+        cand_ensembles = []
+        while(n_clfs < self.max_bag_estimators):
+            new_ens_scores = []
+            for name, new_clf in self.estimators:
+                new_clf_proba = self.clfs_proba_[name]
+
+                new_ens_proba = ((ens_proba * n_clfs + new_clf_proba)
+                                 / (n_clfs + 1))
+                new_ens_score = self.score_metric_(y, new_ens_proba)
+                new_ens_scores.append({'score':new_ens_score,
+                                       'new_estimator_name':name})
+
+            best_ens_score = max(new_ens_scores, key=lambda x: x['score'])
+
+            ens_score = best_ens_score['score']
+            new_estimator_name = best_ens_score['new_estimator_name']
+            ensemble.update({new_estimator_name: 1})
+
+            # store current ensemble to select best later
+            ens_copy = Counter(ensemble)
+            cand = {'ens': ens_copy, 'score': ens_score}
+            cand_ensembles.append(cand)
+
+            n_clfs = sum(ensemble.values())
+
+        if (n_clfs == self.max_bag_estimators):
+            best_cand_ensemble = max(cand_ensembles, key = lambda x: x['score'])
+            ensemble = best_cand_ensemble['ens']
+
+        return ensemble
+
+    def _get_final_ensemble(self, X, y):
+
+        n_clfs = int(len(self.estimators) * (1 - self.prune_fraction))
+        bag_size = int(self.bag_fraction * n_clfs)
+
+        ranked_clfs_score = sorted(self.clfs_score_.items(),
+                                   key=operator.itemgetter(1), reverse=True)
+        ranked_clfs_name = map(lambda (name, clf_score): name,
+                                     ranked_clfs_score)
```

Comment:
weird indentation. In any case you should use a list comprehension here instead. Or `names, scores = zip(*ranked_clf_scores)`